proposal/js-logger.d.ts

```diff
@@ -0,0 +1,206 @@
+   *   * If this warning is associated with a specific span of a Sass
+   *     stylesheet, set `options.span` to a `SourceSpan` that covers that span.
+   *   * If this warning occurred during execution of a stylesheet, set
+   *     `options.stack` to a string representation of the current Sass stack
+   *     trace.
+   *   * Invoke `warn` with a string describing the warning and `options`.
+   *
+   * If this field is defined, the compiler must not surface warnings in any way
+   * other than inokving `warn`.
+   */
+  warn?(
+    message: string,
+    options: {
+      deprecation: boolean;
+      span?: SourceSpan;
+      stack?: string;
+    }
+  ): void;
+
+  /**
+   * If this field is defined, the compiler must invoke it when it encounters a
+   * `@debug` rule using the following procedure:
+   *
+   * * Let `value` be the result of evaluating the rule's expression.
+   * * Let `message` be `value`'s text if it's a string, or the result of
+   *   serializing `value` if it's not.
+   * * Invoke `debug` with `message` and an object with `span` set to the span
+   *   covering the `@debug` rule and its expression.
+   *
+   * If this field is defined, the compiler must not surface debug messages in
+   * any way other than inokving `debug`.
+   */
+  debug?(message: string, options: {span: SourceSpan}): void;
+}
+
+/**
+ * An interface that represents a contiguous section ("span") of a text file.
+ * This section may be empty if the `start` and `end` are the same location,
+ * in which case it indicates a single position in the file.
+ */
+export interface SourceSpan {
+  /**
+   * The location of the first character of this span, unless `end` points to
+   * the same character, in which case the span is empty and refers to the point
+   * between this character and the one before it.
+   */
+  start: SourceLocation;
+
+  /**
+   * The location of the first character after this span. This must point to a
+   * location after `start`.
+   */
+  end: SourceLocation;
+
+  /**
+   * The absolute URL of the file that this span refers to. For files on disk,
+   * this must be a `file://` URL.
+   *
+   * This must be `undefined` for files that are passed to the compiler without
+   * a URL. It must not be `undefined` for any files that are importable.
+   */
+  url?: URL;
+
+  /**
+   * The text covered by the span. This must be the text between `start.offset`
+   * (inclusive) and `end.offset` (exclusive) of the file referred by this
+   * span. Its length must be `end.offset - start.offset`.
+   */
+  text: string;
+
+  /**
+   * Additional source text surrounding this span.
+   *
+   * If this isn't empty, it must contain `text`. Furthermore, `text` must begin
+   * at column `start.column` of a line in `context`.
+   *
+   * > This usually contains the full lines the span begins and ends on if the
+   * > span itself doesn't cover the full lines, but the specific scope is up to
+   * > the compiler.
+   */
+  context: string;
```

Comment:
Would marking this optional be more explicit than letting it be an empty string?

Reply:
Yeah, good point. I think that was a carryover from the embedded protocol.